File: LeetCode/0002-add-two-numbers/0002-add-two-numbers.java
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode root = new ListNode(0);
        ListNode reversed = root;
        int carry = 0;
        
        while (l1 != null || l2 != null || carry != 0) {
            int first = 0;
            int second = 0;
            
            if (l1 != null) {
                first = l1.val;
                l1 = l1.next;
            }
            
            if (l2 != null) {
                second = l2.val;
                l2 = l2.next;
            }
            
            int sum = first + second + carry;
            int added = sum % 10;
            
            reversed.next = new ListNode(added);
            reversed = reversed.next;
            
            carry = sum / 10;
        }
        
        return root.next;
    }
}

File: LeetCode/0002-add-two-numbers/0002-add-two-numbers.kt
class Solution {
    fun addTwoNumbers(l1: ListNode?, l2: ListNode?): ListNode? {
        val l1Num = toBigInteger(reverse(l1))
        val l2Num = toBigInteger(reverse(l2))
        val result = l1Num.add(l2Num)
        
        return toReversedList(result)
    }
    
    fun reverse(head: ListNode?): ListNode? {
        var node = head
        var rev: ListNode? = null
        
        while (node != null) {
            val next = node.next
            node.next = rev
            rev = node
            node = next
        }
        
        return rev
    }
    
    fun toBigInteger(head: ListNode?): BigInteger {
        var node = head
        var str = ""
        
        while (node != null) {
            str += node.`val`
            node = node.next
        }
        
        return BigInteger(str)
    }
    
    fun toReversedList(value: BigInteger): ListNode? {
        var prev: ListNode? = null
        var node: ListNode? = null
        
        for (c in "$value") {
            node = ListNode(Character.getNumericValue(c))
            node.next = prev
            prev = node
        }
        
        return node
    }
}